kube/src/client/mod.rs

```diff
@@ -257,100 +261,51 @@ impl Client {
-                                return Ok(Some((items, (resp, new_buff))));
-                            }
-                        }
-                        Ok(None) => {
-                            trace!("None chunk");
-                            return Ok(None);
+        let frames = FramedRead::new(
+            StreamReader::new(res.into_body().map_err(|e| {
+                if e.is_timeout() {
+                    return std::io::Error::new(std::io::ErrorKind::TimedOut, e);
+                }
+                std::io::Error::new(std::io::ErrorKind::Other, e)
+            })),
+            LinesCodec::new(),
+        );
+
+        Ok(frames.filter_map(|res| async {
+            match res {
+                Ok(line) => match serde_json::from_str::<WatchEvent<T>>(&line) {
+                    Ok(event) => Some(Ok(event)),
+                    Err(e) => {
+                        // Ignore EOF error that can happen for incomplete line from `decode_eof`.
+                        if e.is_eof() {
+                            return None;
                         }
-                        Err(e) => {
-                            if e.is_timeout() {
-                                warn!("timeout in poll: {}", e); // our client timeout
-                                return Ok(None);
-                            }
-                            let inner = e.to_string();
-                            if inner.contains("unexpected EOF during chunk") {
-                                // ^ catches reqwest::Error from hyper::Error
-                                // where the inner.kind == UnexpectedEof
-                                // and the inner.error == "unexpected EOF during chunk size line"
-                                warn!("eof in poll: {}", e);
-                                return Ok(None);
-                            } else {
-                                // There might be other errors worth ignoring here
-                                // For now, if they happen, we hard error up
-                                // This causes a full re-list for Reflector
-                                error!("err poll: {:?} - {}", e, inner);
-                                return Err(Error::HyperError(e));
-                            }
+
+                        // Got general error response
+                        if let Ok(e_resp) = serde_json::from_str::<ErrorResponse>(&line) {
+                            return Some(Err(Error::Api(e_resp)));
                         }
+                        // Parsing error
+                        Some(Err(Error::SerdeError(e)))
+                    }
+                },
+
+                Err(LinesCodecError::Io(e)) => match e.kind() {
+                    // Client timeout
+                    std::io::ErrorKind::TimedOut => {
+                        warn!("timeout in poll: {}", e); // our client timeout
+                        None
                     }
```

Comment:
While this in general seems like a great simplification, we also did use to catch an UnexpectedEof error here (through hyper), which tended to happen after 300+s of watching. I'm not sure how that interacts with this new setup, but it would be good if those errors still pass on a `None`, as it's a restartable error for `watcher`.

Reply:
> which tended to happen after 300+s of watching

I tried watching without a timeout, and the stream finished after about 5 minutes without an error. But this error probably only happens when some event is happening around that time.

I'll map that error with `UnexpectedEof` kind and filter that out.